### PR TITLE
ci: Enable building for s390x and arm64

### PR DESCRIPTION
**Pull request by jschintag**
**What this PR does / why we need it**:
This PR enables building kubesecondarydns for s390x in addition to amd64.
Additionally arm64 is enabled as well, though i can't test it to verify it works.
Additionally it bumps the workflow action versions to their latest to get rid of multiple deprecation warnings.
Finally it fixes a small warning about the case of "AS" in the Dockerfile.

**Special notes for your reviewer**:
PRs to enable unit-tests and build test for s390x will follow. E2E needs the provider, which for s390x still has some problems to be worked out.

### CODE REVIEW

File: Dockerfile
Comment (ashokpariya0) [inline Dockerfile:4]: Can we remove this argument as it's not necessary?

File: Dockerfile
Comment (ashokpariya0) [inline Dockerfile:21]: Can we also add the `ARG TARGETOS` and use it for setting `GOOS`?"

File: Dockerfile
Comment (jschintag) [reply Dockerfile:4]: Theoretically it is needed in the FROM statement, but podman complains if it is stated before it...
I would leave it here to ensure it needs to be set.

File: Dockerfile
Comment (jschintag) [reply Dockerfile:21]: `TARGETOS` does not make much sense, as the outcome we want here is always going to be linux.

Comment (ashokpariya0) [reply Dockerfile:21]: Why? I believe it's better to use the `TARGETOS` value supplied by `buildx` instead of hardcoding it to `linux`, as this Dockerfile wouldn't work for other OS. We can set it like `GOOS=${TARGETOS:-linux}`.

File: Dockerfile
Comment (ashokpariya0) [reply Dockerfile:4]: BUILDPLATFORM is implicitly available when using docker buildx and doesn't need to be explicitly declared using ARG unless you're using it somewhere else in the build process.

File: Dockerfile
Comment (jschintag) [reply Dockerfile:21]: The base image used for the final container is RHEL, so the target is not going to change.
Should it change in the future this can be changed to a variable, but right now it will always be the same value.

File: Dockerfile
Comment (jschintag) [reply Dockerfile:4]: I can drop it sure

File: Dockerfile
Comment (ashokpariya0) [reply Dockerfile:21]: Yes, that makes sense. Thanks!

**Comment by oshoval** [comment]
Once all the comments are ref please let me know and i will review
Thank you all for your contribution

File: Dockerfile
Comment (jschintag) [reply Dockerfile]: Done

**Comment by jschintag** [comment]
@oshoval I have addressed the comments and added the changes from #74 to this PR, as requested in https://github.com/kubevirt/kubesecondarydns/pull/74#issuecomment-2509612317

File: .github/workflows/image-push.yaml
Comment (oshoval) [inline .github/workflows/image-push.yaml:44]: Do we want to add arm please ? (as you added on install go script)
(assuming git actions support arm cross compile)

if you prefer it can wait for follow-ups, unless it is straight forward

Comment (jschintag) [reply .github/workflows/image-push.yaml:44]: Sure i can add it. I just haven't tested it, so if it works or not i can't say.

Comment (jschintag) [reply .github/workflows/image-push.yaml]: I have added it as a separate commit so it is easier to revert should it not work out.

File: Dockerfile
Comment (oshoval) [inline Dockerfile:16]: dont we need something like this ?
https://github.com/kubevirt/bridge-marker/pull/77/files#diff-5aecd73bbcb9b7410938673269f0d4e676dd18509277b0ff2f2787c9d3e22604R3-R11

File: hack/install-go.sh
Comment (oshoval) [reply hack/install-go.sh:22]: lets see please if we do need it imo, because there wasnt an item to support ppc and i prefer to not do just parts unless needed

on the other hand, the suggestion is indeed more compact and robust, so it will be fine to adapt this
but the rest shouldn't be adapted yet

thanks

File: Dockerfile
Comment (jschintag) [reply Dockerfile:16]: Not really, we only need `TARGETARCH` and `ARG` already sets it as an enviroment variable.

File: hack/install-go.sh
Comment (jschintag) [reply hack/install-go.sh:22]: I actually prefer the more verbose version, since it is more easy to read later on and to see what is intended to work.

My proposal would be i keep the switch-case, but change the default to a warning instead of an error.

Comment (oshoval) [reply hack/install-go.sh:22]: i do understand it is more verbose, but it is good to be in the same format on the repos we maintain 
anyhow i wont insist, we can change later according needs

we do need to error imo if one supplied unsupported version in case we keep it this way

Thanks

**Comment by oshoval** [comment]
/lgtm

Thanks, 
Please add arm support to PR desc

**Comment by oshoval** [comment]
/lgtm cancel
/approve

we need another reviewer

File: Dockerfile
Comment (oshoval) [reply Dockerfile:16]: Thanks

@ashokpariya0 does it mean maybe that the logic on bridge-marker or whatever projects that done the same as it can be simplified maybe?

**Comment by oshoval** [comment]
@0xFelix 
can you please take a look ?

**Comment by jschintag** [comment]
> Thanks, Please add arm support to PR desc

done

File: Dockerfile
Comment (0xFelix) [inline Dockerfile:21]: Which image is pulled here? Do the resulting image arch and the BUILDPLATFORM match?

**Review by 0xFelix — COMMENTED** [review]
Did you verify this by running it against a test registry?

File: Dockerfile
Comment (jschintag) [reply Dockerfile:21]: I have tested the whole workflow on my own fork against ghcr.io.
The BUILDPLATFORM means that the first stage is run natively, e.g. on amd64, while the TARGETARCH may be s390x.
The 2nd stage is not affected, so if you target s390x, you compile the s390x binary on amd64 via cross-compile but still get a s390x image in the 2nd stage.

File: Dockerfile
Comment (ashokpariya0) [reply Dockerfile:16]: @oshoval No, it's important to handle default cases. The combination of `ARG` and `ENV` is beneficial because it allows you to manage both build-time and runtime configurations more flexibly and efficiently.

**Review by 0xFelix — APPROVED** [review]
/lgtm

File: Dockerfile
Comment (ashokpariya0) [reply Dockerfile:21]: I believe we need to update the following:
FROM registry.access.redhat.com/ubi8/ubi-minimal
to:
FROM --platform=linux/${TARGETARCH} registry.access.redhat.com/ubi8/ubi-minimal
This is necessary because FROM registry.access.redhat.com/ubi8/ubi-minimal works fine with docker buildx, but it fails with podman when building for multiple platforms.

Comment (0xFelix) [reply Dockerfile:21]: That's what I had in mind.

Comment (oshoval) [reply Dockerfile:21]: Thanks
@jschintag 
would you able to take a look please?

Generally speaking, will be good that we check all the PRs in the effort locally with both docker and podman please

Comment (jschintag) [reply Dockerfile:21]: @oshoval I have tested local builds with both podman and docker. Meaning s390x image build on s390x machine.

I have also tried adding the `--platform` arg to podman, however the result seems to be that it is mostly ignored. Might be that i need to add some extra args or so?

Comment (oshoval) [reply Dockerfile:21]: i am not familiar about it by heart tbh
maybe lets wait please from ideas from the people involved

curious, iiuc it works without it on all the required combinations ? 
native and s390x (cross / native) are most important, on podman mainly
to see it builds and runs

Comment (jschintag) [reply Dockerfile:21]: I have tested native with docker/podman by running `make build` on s390x. Let me test it on x86 as well just to be sure, though there should not be a difference. Cross as i said seemed to not work for me with OCI: podman, host: s390x, target: amd64

Comment (jschintag) [reply Dockerfile:21]: Native amd64 -> amd64 works as well with docker/podman.

Comment (oshoval) [reply Dockerfile:21]: > host: s390x, target: amd64 

we can fix this one later specifically according needs, because atm native host/target amd64 works which is what important

host: amd64, target: s390x - does work ? (podman / docker, higher priority for podman)

maybe the PRs of @ashokpariya0 might help in this, if it works there

Thanks

Comment (ashokpariya0) [reply Dockerfile:21]: The combination of host: s390x and target: amd64 is not supported, so there is no need to test this configuration.
As @oshoval pointed out we need to test following:

Host: amd64
Target: amd64, arm64, and s390x.

Also For the final stage of the build, better to use the following:
FROM --platform=linux/${TARGETARCH} registry.access.redhat.com/ubi8/ubi-minimal
This approach is necessary when building multi-platform images using Podman. You will need to create a Podman manifest and then add the individual platform-specific builds one by one to the manifest. Also adding (linux/${TARGETARCH}) works well with docker buildx.

**Comment by oshoval** [comment]
once we are good to create a release please let me know
thanks

**Comment by jschintag** [comment]
@oshoval I have created https://github.com/kubevirt/kubesecondarydns/pull/78 to address the issue with podman cross-build on amd64. Once that is merged, from my side that should be everything we need for a release of kubesecondarydns.

@ashokpariya0 WDYT?

**Comment by ashokpariya0** [comment]
> @oshoval I have created #78 to address the issue with podman cross-build on amd64. Once that is merged, from my side that should be everything we need for a release of kubesecondarydns.
> 
> @ashokpariya0 WDYT?

>
@jschintag yes sounds good.
@oshoval Looks good to me! Once #78 is merged, we should be all set for the release of kubesecondarydns.

**Comment by oshoval** [comment]
Created tag [v0.0.16](https://github.com/kubevirt/kubesecondarydns/tree/refs/tags/v0.0.16)
It should create a release soon
Thanks all

**Pull request merged**
